### PR TITLE
[IMP] core: sql_db typing

### PR DESCRIPTION
**Pull request by kmagusiak**
Preparation for type checking and some more specific types are used. `PsycoConnection` now is aware of the pool it belongs to.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by rco-odoo — COMMENTED** [review]
I think that this one goes one step too far.
The point of typing Odoo is to help _regular developers_.
No developer should ever use `ConnectionPool`, for instance.  So what's the point of typing implementation details?  This adds no value IMHO.

File: odoo/sql_db.py [hunk 1]
```diff
@@ -140,39 +156,61 @@ def _close(self, rollback):
             super()._close(rollback)
 
 
-class BaseCursor:
+class BaseCursor(_cursor_proxy_type):
```
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/sql_db.py:159]: `BaseCursor` is just a proxy over the psycopg cursor. From the point of view of the user (or typing) it behaves as such. At runtime, it does not inherit from a class to do so, but its implementations overwrite `__getattr__` delegating the actions to the wrapped cursor.

File: odoo/sql_db.py [hunk 2]
```diff
@@ -94,7 +103,12 @@ class Savepoint:
 
     :param BaseCursor cr: the cursor to execute the `SAVEPOINT` queries on
     """
-    def __init__(self, cr):
+    __slots__ = ('_cr', 'closed', 'name')
+    _cr: BaseCursor | psycopg2.extensions.cursor
+    name: str
+    closed: bool
+
+    def __init__(self, cr: BaseCursor | psycopg2.extensions.cursor):
```
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/sql_db.py:111]: We don't use them directly, but `TestCursor` instantiates the savepoint by providing directly the raw cursor.

**Comment by kmagusiak** [comment]
For *regular developers*, when they get an instance of `env.cr`, it is quite empty today from auto-completion perspective.

For us, adding types will allow us to make less mistakes once we setup `mypy` (progressively).

**Review by rco-odoo — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**